src/typecheck.rs

```diff
@@ -1,11 +1,1384 @@
+                    Ok(typ.clone())
+                }
+                (_, Type::NumT(_)) => {
+                    // The inner expression is not a numeric type
+                    Err(TypeError::TypeMismatch {
+                        expected: Type::NumT(0..1), // Example numeric type
+                        actual: inner_type,
+                    })
+                }
+                (Type::NumT(_), _) => {
+                    // The target type is not a numeric type
+                    Err(TypeError::TypeMismatch {
+                        expected: Type::NumT(0..1), // Example numeric type
+                        actual: typ.clone(),
+                    })
+                }
+                _ => {
+                    // Neither the inner expression nor the target type are numeric types
+                    Err(TypeError::TypeMismatch {
+                        expected: Type::NumT(0..1), // Example numeric type
+                        actual: inner_type,
+                    })
+                }
+            }
+        }
+        // Call expressions are always of the type of the function
+        Expr::Call { callee, args } => {
+            // Type check each argument
+            let mut arg_types: Vec<Type> = Vec::new();
+            for arg in args {
+                arg_types.push(typeck_expr(arg, ctx)?);
+            }
+
+            // Look up the function in the function environment
+            let function = ctx
+                .funcs
+                .get(callee)
+                .ok_or(TypeError::UndefinedFunction(*callee))?;
+
+            // Check that the number of arguments matches the number of parameters
+            if arg_types.len() != function.params.len() {
+                return Err(TypeError::WrongNumberOfArguments {
+                    expected: function.params.len(),
+                    actual: arg_types.len(),
+                });
+            }
+
+            // Verify that argument types match parameter types
+            for (arg_type, (_, param_type)) in arg_types.iter().zip(function.params.iter()) {
+                if arg_type != param_type {
+                    return Err(TypeError::TypeMismatch {
+                        expected: param_type.clone(),
+                        actual: arg_type.clone(),
+                    });
+                }
+            }
+
+            // Return the function's return type
+            Ok(function.ret_typ.clone())
+        }
+        // Pattern matching is always of the type of the scrutinee
+        Expr::Match { scrutinee, cases } => {
+            // Type check the scrutinee
+            let scrutinee_type = typeck_expr(scrutinee, ctx)?;
+
+            // If there are no cases, return an error
+            if cases.is_empty() {
+                return Err(TypeError::EmptyMatchCases);
+            }
+
+            // Type check each case and collect result types
+            let mut result_type: Option<Type> = None;
```

Comment:
If you want to maintain the emptyness case above, you can just get the first element and check it to ensure that this is never an `Option<Type>` but just a `Type`.

Something like the following:

```rust
let check_case = |case| { ... }; // the body of the for loop would go here.

let mut cases = cases.iter();

let result_type = check_case(cases.next().unwrap_or_else(|| return Err(EmptyMatchCases)))?;

for case in cases {
  if check_case(case) != result_type { return Err(..); }
}

Ok(result_type)
```